File: ios/RNNetLabelPrinter.m
#import "RNNetLabelPrinter.h"
#import <IOS_SWIFT_WIFI_SDK/IOS_SWIFT_WIFI_SDK-Swift.h>

@implementation RNNetLabelPrinter {
    GTSPL_WIFI *gswifi;
    bool hasListeners;
}

RCT_EXPORT_MODULE()

- (instancetype)init {
    self = [super init];
    if (self) {
        gswifi = [[GTSPL_WIFI alloc] init];
    }
    return self;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"printerStatus"];
}

- (void)startObserving {
    hasListeners = YES;
}

- (void)stopObserving {
    hasListeners = NO;
}

RCT_EXPORT_METHOD(init:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    resolve(nil);
}

RCT_EXPORT_METHOD(openPort:(NSString *)ip
                  port:(NSInteger)port
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi openportWithIP:ip Port:port completion:^(NSString *msg) {
        if ([msg isEqualToString:@"Success"]) {
            resolve(nil);
        } else {
            reject(@"error", msg, nil);
        }
    }];
}

RCT_EXPORT_METHOD(closePort:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi closePort];
    resolve(nil);
}

RCT_EXPORT_METHOD(setup:(NSInteger)width
                  height:(NSInteger)height
                  speed:(NSInteger)speed
                  density:(NSInteger)density
                  sensor:(NSInteger)sensor
                  sensorDistance:(NSInteger)sensorDistance
                  sensorOffset:(NSInteger)sensorOffset
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi setupWithWidth:width height:height speed:speed density:density sensor:sensor sensorDistance:sensorDistance sensorOffset:sensorOffset];
    resolve(nil);
}

RCT_EXPORT_METHOD(clearBuffer:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi clearBuffer];
    resolve(nil);
}

RCT_EXPORT_METHOD(printBarcode:(NSInteger)x
                  y:(NSInteger)y
                  type:(NSString *)type
                  height:(NSInteger)height
                  humanReadable:(NSInteger)humanReadable
                  rotation:(NSInteger)rotation
                  narrow:(NSInteger)narrow
                  wide:(NSInteger)wide
                  content:(NSString *)content
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi printBarcodeWithX:x y:y type:type height:height human_readable:humanReadable rotation:rotation narrow:narrow wide:wide content:content];
    resolve(nil);
}

RCT_EXPORT_METHOD(printFont:(NSInteger)x
                  y:(NSInteger)y
                  fontName:(NSString *)fontName
                  rotation:(NSInteger)rotation
                  xScale:(NSInteger)xScale
                  yScale:(NSInteger)yScale
                  content:(NSString *)content
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi printFontWithX:x y:y fontname:fontName rotation:rotation xScale:xScale yScale:yScale content:content];
    resolve(nil);
}

RCT_EXPORT_METHOD(printFontBlock:(NSInteger)x
                  y:(NSInteger)y
                  width:(NSInteger)width
                  height:(NSInteger)height
                  fontName:(NSString *)fontName
                  rotation:(NSInteger)rotation
                  xScale:(NSInteger)xScale
                  yScale:(NSInteger)yScale
                  space:(NSInteger)space
                  align:(NSInteger)align
                  content:(NSString *)content
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi printFontBlockWithX:x y:y width:width height:height fontname:fontName rotation:rotation xScale:xScale yScale:yScale space:space align:align content:content];
    resolve(nil);
}

RCT_EXPORT_METHOD(printQRCode:(NSInteger)x
                  y:(NSInteger)y
                  eccLevel:(NSString *)eccLevel
                  cellWidth:(NSInteger)cellWidth
                  rotation:(NSInteger)rotation
                  content:(NSString *)content
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi printQRCodeWithX:x y:y eccLevel:eccLevel cellWidth:cellWidth rotation:rotation content:content];
    resolve(nil);
}

RCT_EXPORT_METHOD(printLabel:(NSInteger)set
                  copy:(NSInteger)copy
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    [gswifi printLabelWithSet:set copy:copy];
    resolve(nil);
}

RCT_EXPORT_METHOD(getPrinterStatus:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    NSString *status = [gswifi printerStatus];
    resolve(status);
}

@end 

File: ios/RNNetLabelPrinter.swift
import Foundation
import React
import IOS_SWIFT_WIFI_SDK

@objc(RNNetLabelPrinter)
class RNNetLabelPrinter: RCTEventEmitter {
    private var gswifi: GTSPL_WIFI
    private var hasListeners: Bool = false
    
    override init() {
        gswifi = GTSPL_WIFI()
        super.init()
    }
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return false
    }
    
    @objc
    override func supportedEvents() -> [String] {
        return ["printerStatus"]
    }
    
    @objc
    override func startObserving() {
        hasListeners = true
    }
    
    @objc
    override func stopObserving() {
        hasListeners = false
    }
    
    @objc(init:reject:)
    func init(_ resolve: @escaping RCTPromiseResolveBlock,
              reject: @escaping RCTPromiseRejectBlock) {
        resolve(nil)
    }
    
    @objc(openPort:port:resolve:reject:)
    func openPort(_ ip: String,
                  port: NSNumber,
                  resolve: @escaping RCTPromiseResolveBlock,
                  reject: @escaping RCTPromiseRejectBlock) {
        gswifi.openPort(IP: ip, port: port.intValue) { msg in
            if msg == "connected" {
                resolve(nil)
            } else {
                reject("error", msg, nil)
            }
        }
    }
    
    @objc(closePort:reject:)
    func closePort(_ resolve: @escaping RCTPromiseResolveBlock,
                   reject: @escaping RCTPromiseRejectBlock) {
        gswifi.closePort { msg in
            resolve(nil)
        }
    }
    
    @objc(setup:height:speed:density:sensor:sensorDistance:sensorOffset:resolve:reject:)
    func setup(_ width: NSNumber,
               height: NSNumber,
               speed: NSNumber,
               density: NSNumber,
               sensor: NSNumber,
               sensorDistance: NSNumber,
               sensorOffset: NSNumber,
               resolve: @escaping RCTPromiseResolveBlock,
               reject: @escaping RCTPromiseRejectBlock) {
        gswifi.setup(width: width.intValue,
                    height: height.intValue,
                    speed: speed.intValue,
                    density: density.intValue,
                    sensor: sensor.intValue,
                    sensorDistance: sensorDistance.intValue,
                    sensorOffset: sensorOffset.intValue) { msg in
            if msg == "Connection doesn't exist." {
                reject("connection_error", msg, nil)
            } else {
                resolve(nil)
            }
        }
    }
    
    @objc(clearBuffer:reject:)
    func clearBuffer(_ resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        gswifi.clearBuffer()
        resolve(nil)
    }
    
    @objc(printBarcode:y:type:height:humanReadable:rotation:narrow:wide:content:resolve:reject:)
    func printBarcode(_ x: NSNumber,
                      y: NSNumber,
                      type: String,
                      height: NSNumber,
                      humanReadable: NSNumber,
                      rotation: NSNumber,
                      narrow: NSNumber,
                      wide: NSNumber,
                      content: String,
                      resolve: @escaping RCTPromiseResolveBlock,
                      reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printBarcode(x: x.intValue,
                          y: y.intValue,
                          type: type,
                          height: height.intValue,
                          readable: humanReadable.intValue,
                          rotation: rotation.intValue,
                          narrow: narrow.intValue,
                          wide: wide.intValue,
                          content: content) { msg in
            resolve(nil)
        }
    }
    
    @objc(printFont:y:fontName:rotation:xScale:yScale:content:resolve:reject:)
    func printFont(_ x: NSNumber,
                   y: NSNumber,
                   fontName: String,
                   rotation: NSNumber,
                   xScale: NSNumber,
                   yScale: NSNumber,
                   content: String,
                   resolve: @escaping RCTPromiseResolveBlock,
                   reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printFont(x: x.intValue,
                        y: y.intValue,
                        fontName: fontName,
                        rotation: rotation.intValue,
                        x_scale: xScale.intValue,
                        y_scale: yScale.intValue,
                        content: content) { msg in
            resolve(nil)
        }
    }
    
    @objc(printFontBlock:y:width:height:fontName:rotation:xScale:yScale:space:align:content:resolve:reject:)
    func printFontBlock(_ x: NSNumber,
                        y: NSNumber,
                        width: NSNumber,
                        height: NSNumber,
                        fontName: String,
                        rotation: NSNumber,
                        xScale: NSNumber,
                        yScale: NSNumber,
                        space: NSNumber,
                        align: NSNumber,
                        content: String,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printFontBlock(x: x.intValue,
                            y: y.intValue,
                            width: width.intValue,
                            height: height.intValue,
                            fontName: fontName,
                            rotation: rotation.intValue,
                            x_scale: xScale.intValue,
                            y_scale: yScale.intValue,
                            space: space.intValue,
                            align: align.intValue,
                            content: content) { msg in
            resolve(nil)
        }
    }
    
    @objc(printQRCode:y:eccLevel:cellWidth:rotation:content:resolve:reject:)
    func printQRCode(_ x: NSNumber,
                     y: NSNumber,
                     eccLevel: String,
                     cellWidth: NSNumber,
                     rotation: NSNumber,
                     content: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printQRCode(x: x.intValue,
                          y: y.intValue,
                          eccLevel: eccLevel,
                          cellWidth: cellWidth.intValue,
                          rotation: rotation.intValue,
                          content: content) { msg in
            resolve(nil)
        }
    }
    
    @objc(printLabel:copy:resolve:reject:)
    func printLabel(_ set: NSNumber,
                    copy: NSNumber,
                    resolve: @escaping RCTPromiseResolveBlock,
                    reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printLabel(set: set.intValue, copy: copy.intValue)
        resolve(nil)
    }
    
    @objc(getPrinterStatus:reject:)
    func getPrinterStatus(_ resolve: @escaping RCTPromiseResolveBlock,
                         reject: @escaping RCTPromiseRejectBlock) {
        gswifi.printerStatus { status in
            resolve(status)
        }
    }
    
    @objc(sendCommand:resolve:reject:)
    func sendCommand(_ command: String,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        if command.isEmpty {
            reject("invalid_command", "Command string cannot be empty", nil)
            return
        }
        
        gswifi.sendCommand(command)
        resolve(nil)
    }
    
    @objc(sendByteCmd:resolve:reject:)
    func sendByteCmd(_ byteArray: [NSNumber],
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        if byteArray.isEmpty {
            reject("invalid_data", "Byte array cannot be empty", nil)
            return
        }
        
        let data = byteArray.map { $0.uint8Value }
        gswifi.sendByteCmd(cmdData: Data(data))
        resolve(nil)
    }
} 